### PR TITLE
SpaServices.Extensions: add optional port parameter

### PR DESCRIPTION
**Pull request by stephtr**
Adds an option to specify the port being used by `UseReactDevelopmentServer` or `UseAngularCliServer`.
(converted from [JavaScriptServices/1775](https://github.com/aspnet/JavaScriptServices/pull/1775) )
Addresses #5252

### CODE REVIEW

File: src/Middleware/SpaServices.Extensions/src/AngularCli/AngularCliMiddleware.cs [hunk 1]
```diff
@@ -36,10 +37,25 @@ internal static class AngularCliMiddleware
                 throw new ArgumentException("Cannot be null or empty", nameof(npmScriptName));
             }
 
-            // Start Angular CLI and attach to middleware pipeline
-            var appBuilder = spaBuilder.ApplicationBuilder;
-            var logger = LoggerFinder.GetOrCreateLogger(appBuilder, LogCategoryName);
-            var angularCliServerInfoTask = StartAngularCliServerAsync(sourcePath, npmScriptName, logger);
+            // We have to start the server by ourself if there wasn't any port specified
+            // or the specified port is still free (unused).
+            var shouldStartServer = !spaPort.HasValue || TcpPortFinder.TestPortAvailability(spaPort.Value);
```
Comment (ryanbrandenburg) [inline src/Middleware/SpaServices.Extensions/src/AngularCli/AngularCliMiddleware.cs]: It feels like in the case that a port was specified but it's available we should either

1. Start the server on that port (below starts the server on an arbitrary port).
2. Fail with a message informing you that your server didn't start/was stopped.

File: src/Middleware/SpaServices.Extensions/src/AngularCli/AngularCliMiddlewareOptions.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+{
+    /// <summary>
+    /// Describes options for starting the Angular CLI process
+    /// </summary>
+    public class AngularCliMiddlewareOptions
+    {
+        /// <summary>
+        /// The name of the script in your package.json file that launches the Angular CLI process.
+        /// </summary>
+        public string npmScript;
```
Comment (ryanbrandenburg) [inline src/Middleware/SpaServices.Extensions/src/AngularCli/AngularCliMiddlewareOptions.cs]: Shouldn't this be 
```
public string NpmScript;
```
?
(I don't know why GitHub won't let me make that a suggestion).

File: src/Middleware/SpaServices.Extensions/src/AngularCli/AngularCliMiddlewareOptions.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        /// <summary>
+        /// Port the Angular CLI process should use, if it is not already running.
+        /// </summary>
+        public int? spaPort;
```
Comment (ryanbrandenburg) [inline src/Middleware/SpaServices.Extensions/src/AngularCli/AngularCliMiddlewareOptions.cs]: Casing.

File: src/Middleware/SpaServices.Extensions/src/Util/TcpPortFinder.cs [hunk 4]
```diff
@@ -21,5 +22,19 @@ public static int FindAvailablePort()
                 listener.Stop();
             }
         }
+
+        public static bool TestPortAvailability(int port)
```
Comment (ryanbrandenburg) [inline src/Middleware/SpaServices.Extensions/src/Util/TcpPortFinder.cs]: To me this naming makes it ambiguous what a `true` result means. Maybe `IsPortAvailable`?

File: src/Middleware/SpaServices.Extensions/src/Util/TcpPortFinder.cs [hunk 5]
```diff
@@ -21,5 +22,19 @@ public static int FindAvailablePort()
...
+        {
+            var ipProperties = IPGlobalProperties.GetIPGlobalProperties();
+            var ipEndPoints = ipProperties.GetActiveTcpListeners();
+            foreach (var endPoint in ipEndPoints)
```
Comment (ryanbrandenburg) [inline src/Middleware/SpaServices.Extensions/src/Util/TcpPortFinder.cs]: Nit:
```
return !ipEndpoints.Any(e => e.Port == port);
```

**Review by ryanbrandenburg — CHANGES_REQUESTED** [review]
Seems like the general idea here is correct, just have a couple asks to make things clearer.

File: src/Middleware/SpaServices.Extensions/src/AngularCli/AngularCliMiddleware.cs (hunk 1 above)
Comment (stephtr) [reply src/Middleware/SpaServices.Extensions/src/AngularCli/AngularCliMiddleware.cs]: My use case for this PR is, that I want to be able to either let dotnet start the npm script or start it by myself such that webpack doesn't have to recompile everything, when dotnet has to rebuild some files (and restart the dotnet server). Specifying the port is only relevant for the second case. If the script isn't running and dotnet starts it by itself, I don't care about the port being used.
However to be consistent, approach 1 is in my opinion the best way to go.

**Comment by stephtr** [comment]
Thank you for the useful remarks! Since the original PR was over half a year old, I in the meantime deleted my fork, I therefore have replicated this PR (including your proposed changes) in #12489.

**Pull request closed without merging**